src/librustc_mir/transform/qualify_min_const_fn.rs

```diff
@@ -390,8 +390,12 @@ fn check_terminator(
             cleanup: _,
         } => check_operand(tcx, cond, span, def_id, body),
 
+        | TerminatorKind::FalseUnwind { .. }
+        if feature_allowed(tcx, def_id, sym::const_loop)
+        => Ok(()),
```

Comment:
This seems *very* fragile. `FalseUnwind` is supposed to be ignored (unless you care about unwind paths).

What this code should be doing is a CFG cyclicity check.

Reply:
There is one in [`check_consts`](https://github.com/rust-lang/rust/blob/48840618382eccb8a799320c8e5d08e3b52f4c42/src/librustc_mir/transform/check_consts/validation.rs#L191-L195) that will run even if `#![feature(const_fn)]` is specified. I hope we can get lose that feature gate in favor of fine-grained ones, as well as merge `check_consts` and `qualify_min_const_fn`, in the future.

Reply:
Do *both* run? That makes me happier, I guess, but then the one here is not needed at all, is it?

Reply:
`check_consts` validation is always run. `qualify_min_const_fn` is run when `#![feature(const_fn)]` is not specified. My opinion is that most of this pass should be removed and any check that isn't already in `check_consts` given its own feature gate along with an entry in `check_consts/ops.rs`. I don't wanna spend political capital driving this forward, however.

Reply:
`qualify_min_const_fn` was added as a hack back when the normal const-checking pass was such a mess that nobody was confident in its ability to rule out all bad code. It was always meant to be temporary.

So, assuming const-check these days is well-structured enough that everyone is confident in its ability to do what it is supposed to do, I don't think you'll need to spend any political capital to get rid of it. Quite the contrary, at least I personally would be delighted to see it go. :)

Reply:
Yes, `qualify_min_const_fn` should go.

Reply:
Ah, my bad. I didn't know that there was consensus that the new const checker is sufficiently unmessy.

Reply:
I would prefer if we could wait with removing `qualify_min_const_fn`. Not necessarily because the new const checker is messy, but because I don't understand the new code (which is based on data-flow) nearly as well as the old code. And as the lang team point-person for the const things, I would like to understand the thing that checks for stability wrt. const. So I'd appreciate if we don't do it right now and allow me some time to study the new stuff. (Also, is there an urgency to this?)

Reply:
There's no urgency, there are some preliminary things that can be done like eliminating the `const_fn` feature gate

Reply:
You could just make the handling of `FalseUnwind` consistent.
It's not, and should not be mistaken for, a loop.

As the name says, it's a "false" (or "phantom") unwind edge. It's for situations where analyses which care about unwinding / cleanup blocks need to be conservative and assume unwinding can happen "out of the blue".

Everything else in const-checking ignores unwind edges (e.g. from calls) and associated cleanup blocks, because we completely bypass unwinding at compile-time.

Reply:
See #62274 for precedent.

Reply:
Also looks like `FalseEdges` is also mishandled?